### PR TITLE
all: invenio version 3.2 and availability

### PR DESCRIPTION
**Pull request by rerowep**
* Updates invenio to version 3.2.
* Bumps ES to version 7
* Adds XSL style to OAI2D.
* Implemnets deletetion of resource links for unavailable ebooks during harvesting.

### CODE REVIEW

**Comment by rerowep** [comment]
> Commit proposition:
> 
> ```
> all: update to invenio 3.2, clean on availability status
> 
> * Updates invenio to version 3.2.
> * Bumps ES to version 7
> * Adds XSL style to OAI2D.
> * Implements deletion of resource links for unavailable ebooks during harvesting.
> ```
> 
> Also, in the future, even for `rero-ebooks` of MEF, try not to squeeze everything in the same commit. Maybe it's better to have two clear commit in the same PR.

The first line is longer then 50 characters.

**Comment by iGormilhit** [comment]
> > Commit proposition:
> > ```
> > all: update to invenio 3.2, clean on availability status
> > 
> > * Updates invenio to version 3.2.
> > * Bumps ES to version 7
> > * Adds XSL style to OAI2D.
> > * Implements deletion of resource links for unavailable ebooks during harvesting.
> > ```
> > 
> > 
> > Also, in the future, even for `rero-ebooks` of MEF, try not to squeeze everything in the same commit. Maybe it's better to have two clear commit in the same PR.
> 
> The first line is longer then 50 characters.

Ok, that's a propostion. Try to begin with a verb. With a commit by feature, that's easier. :wink: 

But for the list, the typos are still there, I have corrected them in my proposition.

File: docker-services.yml [hunk 1]
```diff
@@ -64,7 +64,7 @@ services:
     ports:
       - "6379:6379"
   db:
-    image: postgres:9.6
+    image: postgres:12
```
Comment (jma) [inline docker-services.yml:67]: Why do you need this? What is the advantage of this new version. Moreover you do not have any mention in you commit message.

File: docker/nginx/nginx.conf [hunk 2]
```diff
@@ -36,11 +36,16 @@ http {
     include       /etc/nginx/mime.types;
     default_type  application/octet-stream;
 
+    map "$request:$status:$body_bytes_sent" $loggable {
```
Comment (jma) [inline docker/nginx/nginx.conf:39]: Cool can we do this in rero-ils?

File: docs/conf.py [hunk 3]
```diff
@@ -29,8 +29,6 @@
 
 import os
 
-import sphinx.environment
```
Comment (jma) [inline docs/conf.py:32]: why?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,99 @@
+[tool.poetry]
+name = "rero-ebooks"
+version = "0.2.0"
+description = "Invenio digital library framework."
```
Comment (jma) [inline pyproject.toml:4]: please change it

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+## Additionnal constraints on python modules
+# solves datetime serialize error
+# celery = "<4.3.0"
+
+## Deployment
+# sentry
+invenio-logging = { version = ">=1.2.0,<1.3.0", extras = ["sentry-sdk", "sentry"] }
+lxml = ">=3.5.0,<4.2.6"
+python-dotenv = "^0.13.0"
+
```
Comment (jma) [inline pyproject.toml:62]: why not uwsgi?

File: rero_ebooks/api.py [hunk 6]
```diff
@@ -31,14 +31,32 @@
...
 
+class EbookError:
+    """Base class for errors in the Ebook class."""
+
+    class PidMissing(Exception):
+        """Ebook pid missing."""
+
+
+class EbooksSearch(RecordsSearch):
+    """EbooksSearch."""
```
Comment (jma) [inline rero_ebooks/api.py:52]: please change it

File: rero_ebooks/api.py [hunk 7]
```diff
@@ -49,31 +67,86 @@ class Ebook(Record):
...
+            for key, value in e_res.items():
+                if isinstance(value, tuple):
+                    e_res[key] = list(value)
+
+        old_e_res = old_record.get(field)
+        for e_res in old_e_res:
+            # check if already exists!
+            if e_res not in new_e_res:
+                new_e_res.append(copy.deepcopy(e_res))
+                idx = new_record['__order__'].index(field)
```
Comment (jma) [inline rero_ebooks/api.py:86]: a small comment is welcome here.

File: rero_ebooks/api.py [hunk 8]
```diff
@@ -49,31 +67,86 @@ class Ebook(Record):
...
+                new_record['__order__'].insert(idx, field)
+        return new_record
+
+    @classmethod
+    def create_or_update(cls, data, id_=None, dbcommit=False, reindex=False,
+                         vendor=None, **kwargs):
```
Comment (jma) [inline rero_ebooks/api.py:92]: please describe the input parameters in the docstring. Please check also other methods.

File: rero_ebooks/api.py [hunk 9]
```diff
@@ -96,71 +169,61 @@ def create(
...
-            try:
-                RecordIndexer().delete(record)
-            except NotFoundError:
-                pass
-        return result
+            self.delete_from_index()
+        return self
+
+    @classmethod
+    def get_record_by_id(cls, id, with_deleted=False):
```
Comment (jma) [inline rero_ebooks/api.py:184]: useless

File: rero_ebooks/apiharvester/cli.py [hunk 10]
```diff
@@ -108,6 +108,22 @@ def api_source_config_from_file(configfile, update):
...
 
+@apiharvester.command('initsets')
+@click.argument('configfile', type=click.File('rb'))
+@click.option('-v', '--verbose', 'verbose', is_flag=True, default=False)
+@with_appcontext
+def init_oai_sets(configfile, verbose):
+    """Init OAIsets."""
+    configs = yaml.load(configfile, Loader=yaml.FullLoader)
+    for name, values in sorted(configs.items()):
+        description = values.get('description', '...')
```
Comment (jma) [inline rero_ebooks/apiharvester/cli.py:119]: why ...?

File: rero_ebooks/apiharvester/utils.py [hunk 11]
```diff
@@ -28,10 +28,31 @@
...
 from invenio_db import db
+from invenio_oaiserver.models import OAISet
 
 from .models import ApiHarvestConfig
 
 
+def add_set(spec, name, pattern, description='...'):
+    """Add OAI set.
+
+    :param spec: set identifier
```
Comment (jma) [inline rero_ebooks/apiharvester/utils.py:39]: Great!

File: rero_ebooks/cantook/api.py [hunk 12]
```diff
@@ -76,7 +82,7 @@ def get_request_url(self, start_date='1990-01-01', page=1):
             start_date=start_date,
             page=page
         )
-        return self.config.url + '/v1/resources.json?{params}'.format(
+        return self._url + '/v1/resources.json?{params}'.format(
```
Comment (jma) [inline rero_ebooks/cantook/api.py:85]: can you rewrite without `+`?

File: rero_ebooks/cantook/api.py [hunk 13]
```diff
@@ -113,21 +119,64 @@ def create_update_record(self, record):
...
-                    click.echo(
-                        '{count}: {pid} {msg}'.format(
-                            count=self._count,
-                            pid=record['pid'],
-                            msg=msg
-                        )
-                    )
+                # get the nature of the media,
+                # because we only like to import epub's
+                nature = 'unknown'
```
Comment (jma) [inline rero_ebooks/cantook/api.py:151]: I would use media_type instead of nature.

File: rero_ebooks/cantook/api.py [hunk 14]
```diff
@@ -136,6 +185,44 @@ def verbose_print(self, msg):
         if self.verbose:
             click.echo(msg)
 
+    def init_available_ids(self, from_date):
+        """Get all aavailable pids.
+
+        from_date: record changed after this date to get
```
Comment (jma) [inline rero_ebooks/cantook/api.py:191]: missing keyword param...

File: rero_ebooks/cantook/api.py [hunk 15]
```diff
@@ -136,6 +185,44 @@ def verbose_print(self, msg):
...
+        """
+        url = self.get_request_url(start_date=from_date, page=1)
+        url += '&available=1'
```
Comment (jma) [inline rero_ebooks/cantook/api.py:194]: please use format

File: rero_ebooks/cantook/api.py [hunk 16]
```diff
@@ -136,6 +185,44 @@ def verbose_print(self, msg):
...
+            )
+            for record in request.json().get('resources', []):
+                count += 1
+                self._available_ids[record['id']] = count
+            # get next page and update current_page
+            url = self.get_request_url(
+                start_date=from_date,
+                page=current_page+1
+            )
+            url += '&available=1'
```
Comment (jma) [inline rero_ebooks/cantook/api.py:218]: please use format

File: rero_ebooks/cantook/api.py [hunk 17]
```diff
@@ -171,10 +259,36 @@ def get_records(self, from_date, max=0, file=None):
...
+        return self._count_new
+
+    @property
+    def count_upd(self):
+        """Get updated count."""
+        return self._count_upd
+
+    @property
+    def count_del(self):
+        """Get delteed count."""
```
Comment (jma) [inline rero_ebooks/cantook/api.py:288]: deleted...

File: rero_ebooks/config.py [hunk 18]
```diff
@@ -47,8 +45,10 @@ def _(x):
 
 # Rate limiting
 # =============
-#: Storage for ratelimiter.
 RATELIMIT_STORAGE_URL = 'redis://localhost:6379/3'
+RATELIMIT_DEFAULT = '5000/second'
+RATELIMIT_ENABLED = False
```
Comment (jma) [inline rero_ebooks/config.py:50]: does it make sense to have a default rate limit if it is disabled?

File: rero_ebooks/ext.py [hunk 19]
```diff
@@ -54,4 +50,3 @@ def init_config(self, app):
     @staticmethod
     def register_signals(app):
```
Comment (jma) [inline rero_ebooks/ext.py:51]: please remove this method.

File: rero_ebooks/ext.py [hunk 20]
```diff
@@ -54,4 +50,3 @@ def init_config(self, app):
...
         """Register Zenodo Deposit signals."""
```
Comment (jma) [inline rero_ebooks/ext.py:52]: ?? zenodo??

File: rero_ebooks/mappings/v7/ebooks/ebook-v1.0.0.json [hunk 21]
```diff
@@ -0,0 +1,7976 @@
+{
+  "settings": {
```
Comment (jma) [inline rero_ebooks/mappings/v7/ebooks/ebook-v1.0.0.json:2]: what do we need to be indexed in this instance?

File: docker-services.yml (hunk 1 above)
Comment (rerowep) [reply docker-services.yml:67]: To be up to date with postgress.

File: docker/nginx/nginx.conf (hunk 2 above)
Comment (rerowep) [reply docker/nginx/nginx.conf:39]: Its a long time waiting in https://github.com/rero/rero-ils/pull/677

File: docs/conf.py (hunk 3 above)
Comment (rerowep) [reply docs/conf.py:32]: `autoflake` found this as unused import.

File: pyproject.toml (hunk 4 above)
Comment (rerowep) [reply pyproject.toml:4]: I changed the version to `0.3.0` and the description to `Ebooks repository for RERO`

File: rero_ebooks/config.py (hunk 18 above)
Comment (rerowep) [reply rero_ebooks/config.py:50]: Yes if we enable it in deployment with an environment variable.

File: rero_ebooks/ext.py (hunk 20 above)
Comment (rerowep) [reply rero_ebooks/ext.py:52]: Old code, I have changed to: `Register Ebooks signals.`

File: rero_ebooks/mappings/v7/ebooks/ebook-v1.0.0.json (hunk 21 above)
Comment (rerowep) [reply rero_ebooks/mappings/v7/ebooks/ebook-v1.0.0.json:2]: This is the standard marc21 mapping from invenio. Mappings are necessary for declaring `OAI sets`, searching in the API and search in the OAI requests.

File: rero_ebooks/ext.py (hunk 19 above)
Comment (rerowep) [reply rero_ebooks/ext.py:51]: ok

File: rero_ebooks/cantook/api.py (hunk 17 above)
Comment (rerowep) [reply rero_ebooks/cantook/api.py:288]: ok

File: pyproject.toml (hunk 5 above)
Comment (rerowep) [reply pyproject.toml:62]: This file is more or less a copy from `rero-ils` and `uwsgi` is defined in line 43.

**Pull request merged**